ocaml/forkexecd/lib/fe_stubs.c

```diff
@@ -0,0 +1,405 @@
+        execve(args[0], args, envs);
+        // keep compatibility with forkexecd daemon.
+        _exit(errno == ENOENT ? 127 : 126);
+    }
+
+    // Restore thread state
+    pthread_sigmask(SIG_SETMASK, &old_sigset, NULL);
+    pthread_setcancelstate(cancellation_state, NULL);
+
+    // We don't need writing pipe anymore and we need to detect
+    // if closed so we can't keep it open
+    close_fd(&pipe_fds[1]);
+
+    if (err != 0) {
+        close_fd(&pipe_fds[0]);
+        res->err_msg = "vfork";
+        return err;
+    }
+
+    // Handle errors from helper
+    if (pipe_fds[0] >= 0) {
+        int readed;
+        // Note that buffer is small and written atomically by
+        // the helper, no reason for the kernel to split it.
+        while ((readed = read(pipe_fds[0], &res->msg, sizeof(res->msg))) < 0
+               && errno == EINTR)
+            continue;
+        close_fd(&pipe_fds[0]);
+        if (readed != 0 && readed < offsetof(msg_t, msg_buf) + 1) {
+            // This should never happen !!!
+            // At this point the process is created and we have a pid so
+            // we cannot just return an error.
+            // We could try to wait the process but it should fail, let
+            // returns success and let caller read process status result.
+            return 0;
+        }
+        res->msg.msg_buf[sizeof(res->msg.msg_buf) - 1] = 0;
+        if (readed > 0) {
+            // Wait the process otherwise we'll have a zombie
+            reap_pid(res->pid);
+
+            res->err_msg = res->msg.msg_buf;
+            return res->msg.err;
+        }
+    }
+    return 0;
+}
+
+CAMLprim value
+caml_safe_exec_with_helper(value args, value environment)
+{
+    CAMLparam2(args, environment);
+
+    // Copy parameters to C
+    char **c_args = copy_string_list(args);
+    char **c_envs = copy_string_list(environment);
+    if (!c_envs || !c_args) {
+        free(c_envs);
+        free(c_args);
+        caml_raise_out_of_memory();
+    }
+
+    // potentially slow section, release Ocaml engine
+    caml_enter_blocking_section();
+
+    safe_exec_result res;
+    int err = safe_exec_with_helper(&res, c_args, c_envs);
+
+    free(c_envs);
+    free(c_args);
+
+    caml_leave_blocking_section();
+
+    // error, notify with an exception
+    if (err != 0)
+        unix_error(err, res.err_msg, Nothing);
+
+    CAMLreturn(Val_int(res.pid));
+}
+
+CAMLprim value
+caml_pidwaiter_dontwait(value pid_val)
+{
+    CAMLparam1(pid_val);
+    pid_t pid = Int_val(pid_val);
+
+    // reap the pid to avoid zombies
+    pthread_t th;
+    if (create_thread_minstack(&th, thread_proc_reap, (void *) (intptr_t) pid) == 0)
+        pthread_detach(th);
+
+    CAMLreturn(Val_unit);
+}
+
+typedef struct {
+    pid_t pid;
+    bool timed_out;
+    bool stop;
+    struct timespec deadline;
+    pthread_mutex_t mtx;
+    pthread_cond_t cond;
+} timeout_kill;
+
+static void *
+thread_proc_timeout_kill(void *arg)
```

Comment:
XAPI already has a singleshot timer where we can register something to happen after a given time (in OCaml), so we don't necessarily need to spawn a separate thread to do that.
See 'make_timeboxed_rpc' for an example on how to use it.

Reply:
This looks a pretty regression issue on the design. Currently, this library is pretty independent, being able to test it separately. The code you pointed out is using some Xapi code that is really Xapi dependent, requiring some thread to be spawn.

We'll discuss Wednesday about.

Reply:
To sum up discussion, we are rewriting this part in Ocaml trying to reuse Xapi scheduler (making it a separate library).